app/scripts/migrations/139.ts

```diff
@@ -0,0 +1,161 @@
+ * controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(originalVersionedData: {
+  meta: { version: number };
+  data: Record<string, unknown>;
+}) {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  versionedData.data = transformState(versionedData.data);
+  return versionedData;
+}
+
+/**
+ * Transform the MetaMask extension state to add the
+ * `endowment:permitted-chains` permission to all Snaps that have the
+ * `endowment:ethereum-provider` permission.
+ *
+ * If the `PermissionController` or `SelectedNetworkController` state is not
+ * found or is not an object, the state is returned as-is.
+ *
+ * @param state - The MetaMask extension state.
+ * @returns The updated MetaMask extension state.
+ */
+function transformState(state: Record<string, unknown>) {
+  const {
+    PermissionController: permissionControllerState,
+    SelectedNetworkController: selectedNetworkControllerState,
+  } = state;
+  if (!permissionControllerState || !isObject(permissionControllerState)) {
+    log.warn(
+      'Skipping migration: `PermissionController` state not found or is not an object.',
+    );
+    return state;
+  }
+
+  if (!isObject(permissionControllerState.subjects)) {
+    log.warn(
+      'Skipping migration: `PermissionController.subjects` state is not an object.',
+    );
+    return state;
+  }
+
+  if (
+    !selectedNetworkControllerState ||
+    !isObject(selectedNetworkControllerState)
+  ) {
+    log.warn(
+      'Skipping migration: `SelectedNetworkController` state not found or is not an object.',
+    );
+    return state;
+  }
+
+  if (!isObject(selectedNetworkControllerState.domains)) {
+    log.warn(
+      'Skipping migration: `SelectedNetworkController.domains` state is not an object.',
+    );
+    return state;
+  }
+
+  const currentChainId =
+    selectedNetworkControllerState.domains[METAMASK_DOMAIN];
+
+  const updatedSubjects: string[] = [];
+
+  // Add permission to use the current globally selected network to all Snaps
+  // that have the `endowment:ethereum-provider` permission.
+  const entries = Object.entries(permissionControllerState.subjects) as [
+    string,
+    GenericPermissionControllerSubject,
+  ][];
+
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration =
+      permissionKeys.includes(SnapEndowments.EthereumProvider) &&
+      !permissionKeys.includes(PermissionNames.permittedChains);
+
+    if (!needsMigration) {
+      return {
+        ...accumulator,
+        [key]: subject,
+      };
+    }
+
+    updatedSubjects.push(key);
+
+    const newSubject: GenericPermissionControllerSubject = {
+      ...subject,
+      permissions: {
+        ...subject.permissions,
+        [PermissionNames.permittedChains]: {
```

Comment:
This permission has been removed in favor of CAIP-25, so we'll need to adjust this slightly